File: android/app/src/main/java/com/bluedrop/bluetooth/BdipSession.kt
package com.bluedrop.bluetooth

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull
import java.io.InputStream
import java.io.OutputStream
import java.util.concurrent.atomic.AtomicBoolean

/**
 * One BDIP session over an established socket (see docs/PROTOCOL.md).
 *
 * Transport-agnostic (plain streams + a close callback) so it can be unit
 * tested with in-memory pipes. All writes are serialized through a single
 * writer coroutine, as the spec requires.
 */
class BdipSession(
    private val scope: CoroutineScope,
    private val input: InputStream,
    private val output: OutputStream,
    private val closeTransport: () -> Unit,
    private val listener: Listener,
    private val hello: HelloInfo,
    private val role: Role,
) {
    enum class Role { INITIATOR, RESPONDER }

    interface Listener {
        fun onEstablished(peer: HelloInfo)
        fun onClipboardText(text: String)
        fun onClipboardImage(png: ByteArray)
        fun onFileMeta(meta: FileMeta)
        fun onFileChunk(index: Int, data: ByteArray)
        fun onFileAck(ack: FileAck)
        fun onClosed(reason: String)
    }

    private val writeChannel = Channel<ByteArray>(Channel.UNLIMITED)
    private val ackChannel = Channel<FileAck>(Channel.UNLIMITED)
    private val pendingHello = Channel<Frame>(Channel.UNLIMITED)
    private val closed = AtomicBoolean(false)
    private val established = CompletableDeferred<Unit>()

    @Volatile private var lastSentAt = 0L
    @Volatile private var lastReceivedAt = 0L
    private var readerJob: Job? = null
    private var writerJob: Job? = null
    private var heartbeatJob: Job? = null
    private var handshakeJob: Job? = null

    /** Per-chunk ACK wait; overridable so tests can shorten it. */
    var chunkAckTimeoutMs: Long = 60_000

    /** HELLO exchange timeout; overridable so tests can extend it on slow runners. */
    var helloTimeoutMs: Long = Bdip.HELLO_TIMEOUT_MS

    val isEstablished: Boolean get() = established.isCompleted

    // ---------------------------------------------------------------- lifecycle

    fun start() {
        val now = System.currentTimeMillis()
        lastSentAt = now
        lastReceivedAt = now
        writerJob = scope.launch {
            for (bytes in writeChannel) {
                output.write(bytes)
                output.flush()
                lastSentAt = System.currentTimeMillis()
            }
        }
        readerJob = scope.launch { runReader() }
        heartbeatJob = scope.launch { runHeartbeat() }
        handshakeJob = scope.launch { runHandshake() }
    }

    fun shutdown(detail: String = "") {
        sendBye(Bdip.BYE_SHUTDOWN, detail)
        tearDown("bye sent")
    }

    private fun fail(message: String) {
        sendBye(Bdip.BYE_PROTOCOL_ERROR, message)
        tearDown(message)
    }

    private fun sendBye(reason: Int, detail: String) {
        try {
            val payload = byteArrayOf(reason.toByte()) +
                ("{\"message\":\"$detail\"}").toByteArray(Charsets.UTF_8)
            writeChannel.trySend(FrameCodec.encode(Bdip.TYPE_BYE, payload))
            // give the writer a moment to flush the BYE before hard close
            Thread.sleep(150)
        } catch (_: Exception) {
        }
    }

    private fun tearDown(reason: String) {
        if (!closed.compareAndSet(false, true)) return
        readerJob?.cancel()
        writerJob?.cancel()
        heartbeatJob?.cancel()
        handshakeJob?.cancel()
        writeChannel.close()
        ackChannel.close()
        pendingHello.close()
        established.cancel()
        try {
            closeTransport()
        } catch (_: Exception) {
        }
        listener.onClosed(reason)
    }

    // ---------------------------------------------------------------- handshake

    private suspend fun runHandshake() {
        try {
            if (role == Role.INITIATOR) {
                writeChannel.send(FrameCodec.encode(Bdip.TYPE_HELLO, hello.encode()))
                lastSentAt = System.currentTimeMillis()
            }
            val peer = withTimeoutOrNull(helloTimeoutMs) { pendingHello.receive() }
            if (peer == null || peer.type != Bdip.TYPE_HELLO) {
                fail("hello timeout")
                return
            }
            val info = try {
                HelloInfo.decode(peer.payload)
            } catch (e: Exception) {
                fail("bad HELLO payload: ${e.message}")
                return
            }
            if (hello.token.isNotEmpty() && info.token.isNotEmpty() && hello.token != info.token) {
                fail("pairing token mismatch")
                return
            }
            if (role == Role.RESPONDER) {
                writeChannel.send(FrameCodec.encode(Bdip.TYPE_HELLO, hello.encode()))
                lastSentAt = System.currentTimeMillis()
            }
            established.complete(Unit)
            listener.onEstablished(info)
        } catch (e: Exception) {
            if (!closed.get()) fail("handshake failed: ${e.message}")
        }
    }

    // ---------------------------------------------------------------- loops

    private suspend fun runReader() {
        try {
            while (!closed.get()) {
                val frame = FrameCodec.read(input)
                lastReceivedAt = System.currentTimeMillis()
                if (!established.isCompleted) {
                    when (frame.type) {
                        Bdip.TYPE_HELLO -> {
                            pendingHello.send(frame)
                            continue
                        }

                        Bdip.TYPE_BYE -> {
                            val reason = frame.payload.getOrNull(0)?.toInt() ?: 0
                            tearDown("peer bye ($reason)")
                            return
                        }

                        else -> {
                            fail("expected HELLO first, got type 0x${frame.type.toString(16)}")
                            return
                        }
                    }
                }
                when (frame.type) {
                    Bdip.TYPE_HELLO -> { /* duplicate HELLO after establish: ignore */ }
                    Bdip.TYPE_PING -> writeChannel.send(
                        FrameCodec.encode(Bdip.TYPE_PONG, frame.payload)
                    )

                    Bdip.TYPE_BYE -> {
                        val reason = frame.payload.getOrNull(0)?.toInt() ?: 0
                        tearDown("peer bye ($reason)")
                        return
                    }

                    Bdip.TYPE_CLIPBOARD_TEXT ->
                        listener.onClipboardText(String(frame.payload, Charsets.UTF_8))

                    Bdip.TYPE_CLIPBOARD_IMAGE -> listener.onClipboardImage(frame.payload)

                    Bdip.TYPE_FILE_META -> runCatching { FileMeta.decode(frame.payload) }
                        .onSuccess { listener.onFileMeta(it) }

                    Bdip.TYPE_FILE_CHUNK -> if (frame.payload.size >= 4) {
                        val index = (frame.payload[0].toInt() and 0xFF) or
                            ((frame.payload[1].toInt() and 0xFF) shl 8) or
                            ((frame.payload[2].toInt() and 0xFF) shl 16) or
                            ((frame.payload[3].toInt() and 0xFF) shl 24)
                        listener.onFileChunk(
                            index,
                            frame.payload.copyOfRange(4, frame.payload.size),
                        )
                    }

                    Bdip.TYPE_FILE_ACK -> runCatching { FileAck.decode(frame.payload) }
                        .onSuccess {
                            ackChannel.send(it)
                            listener.onFileAck(it)
                        }

                    else -> { /* unknown type: skip per spec */ }
                }
            }
        } catch (e: Bdip.ProtocolException) {
            if (!closed.get()) fail(e.message ?: "protocol error")
        } catch (e: Exception) {
            if (!closed.get()) tearDown("link lost: ${e.message ?: e.javaClass.simpleName}")
        }
    }

    private suspend fun runHeartbeat() {
        while (!closed.get()) {
            delay(1_000)
            if (closed.get()) return
            val now = System.currentTimeMillis()
            if (now - lastReceivedAt >= Bdip.RX_DEAD_MS) {
                tearDown("rx dead after ${Bdip.RX_DEAD_MS} ms")
                return
            }
            if (now - lastSentAt >= Bdip.PING_INTERVAL_MS) {
                writeChannel.send(
                    FrameCodec.encode(Bdip.TYPE_PING, ByteArray(8) { it.toByte() })
                )
                lastSentAt = now
            }
        }
    }

    // ---------------------------------------------------------------- senders

    /** Sends clipboard text; refuses payloads over the send cap. */
    fun sendClipboardText(text: String): Boolean {
        val bytes = text.toByteArray(Charsets.UTF_8)
        if (bytes.size > Bdip.MAX_TEXT) return false
        return writeChannel.trySend(FrameCodec.encode(Bdip.TYPE_CLIPBOARD_TEXT, bytes)).isSuccess
    }

    fun sendClipboardImage(png: ByteArray): Boolean {
        if (png.size > Bdip.MAX_IMAGE) return false
        return writeChannel.trySend(FrameCodec.encode(Bdip.TYPE_CLIPBOARD_IMAGE, png)).isSuccess
    }

    suspend fun awaitEstablished() = established.await()

    /**
     * Streams a file using stop-and-wait chunking (spec §3.4). Returns the
     * number of bytes acknowledged, or -1 on failure.
     */
    suspend fun sendFile(
        meta: FileMeta,
        readChunk: (offset: Long, count: Int) -> ByteArray,
        onProgress: ((received: Long, total: Long) -> Unit)? = null,
    ): Long {
        writeChannel.send(FrameCodec.encode(Bdip.TYPE_FILE_META, meta.encode()))
        var offset = 0L
        var index = 0
        while (offset < meta.size) {
            val count = minOf(Bdip.CHUNK_SIZE.toLong(), meta.size - offset).toInt()
            val chunk = readChunk(offset, count)
            if (chunk.size != count) return -1
            writeChannel.send(
                FrameCodec.encode(Bdip.TYPE_FILE_CHUNK, FrameCodec.u32(index) + chunk)
            )
            val ack = withTimeoutOrNull(chunkAckTimeoutMs) { ackChannel.receive() } ?: return -1
            if (ack.id != meta.id || ack.error != null) return -1
            offset += count
            index++
            onProgress?.invoke(offset, meta.size)
            if (ack.done == true) return meta.size // receiver combined the final ack
        }
        // the receiver's final ack carries done=true
        val done = withTimeoutOrNull(chunkAckTimeoutMs) { ackChannel.receive() } ?: return -1
        if (done.done != true || done.error != null) return -1
        return meta.size
    }

    fun sendFileAck(ack: FileAck) {
        writeChannel.trySend(FrameCodec.encode(Bdip.TYPE_FILE_ACK, ack.encode()))
    }
}

File: windows/Protocol/BdipSession.cs
using System.Buffers.Binary;
using System.Collections.Concurrent;
using System.IO;

namespace BluedropWindows.Protocol;

public enum BdipRole { Initiator, Responder }

public interface IBdipListener
{
    void OnEstablished(HelloInfo peer);
    void OnClipboardText(string text);
    void OnClipboardImage(byte[] png);
    void OnFileMeta(FileMeta meta);
    void OnFileChunk(int index, byte[] data);
    void OnFileAck(FileAck ack);
    void OnClosed(string reason);
}

/// <summary>
/// One BDIP session over an established stream (see docs/PROTOCOL.md).
/// Writes are serialized through a single consumer loop; the reader
/// dispatches frames to the listener. Mirrors the Kotlin BdipSession.
/// </summary>
public sealed class BdipSession : IDisposable
{
    private readonly Stream _stream;
    private readonly Action _closeTransport;
    private readonly IBdipListener _listener;
    private readonly HelloInfo _hello;
    private readonly BdipRole _role;
    private readonly CancellationTokenSource _cts = new();
    private readonly BlockingCollection<byte[]> _writeQueue = new(boundedCapacity: 1024);
    private readonly ConcurrentQueue<TaskCompletionSource<FileAck>> _ackWaiters = new();
    private readonly TaskCompletionSource<Frame> _peerHello = new();
    private readonly TaskCompletionSource _established = new();

    private volatile bool _closed;
    private long _lastSentAt;
    private long _lastReceivedAt;
    private Task? _readerTask;
    private Task? _writerTask;
    private Task? _heartbeatTask;

    public int ChunkAckTimeoutMs { get; set; } = 60_000;
    public bool IsEstablished => _established.Task.IsCompleted;

    public BdipSession(Stream stream, Action closeTransport,
        IBdipListener listener, HelloInfo hello, BdipRole role)
    {
        _stream = stream;
        _closeTransport = closeTransport;
        _listener = listener;
        _hello = hello;
        _role = role;
    }

    public void Start()
    {
        var now = Environment.TickCount64;
        _lastSentAt = now;
        _lastReceivedAt = now;
        _writerTask = Task.Run(WriterLoop);
        _readerTask = Task.Run(ReaderLoop);
        _heartbeatTask = Task.Run(HeartbeatLoop);
        _ = Task.Run(HandshakeAsync);
    }

    public Task AwaitEstablishedAsync(CancellationToken ct = default) =>
        _established.Task.WaitAsync(ct);

    // ------------------------------------------------------------- lifecycle

    public void Shutdown(string detail = "")
    {
        SendBye(Bdip.ByeShutdown, detail);
        TearDown("bye sent");
    }

    private void Fail(string message)
    {
        SendBye(Bdip.ByeProtocolError, message);
        TearDown(message);
    }

    private void SendBye(byte reason, string detail)
    {
        try
        {
            _writeQueue.TryAdd(new Frame(Bdip.TypeBye, ByeDetail.Encode(reason, detail)).Encode());
            // let the writer flush before the hard close
            Thread.Sleep(150);
        }
        catch { /* already closing */ }
    }

    private void TearDown(string reason)
    {
        if (_closed) return;
        _closed = true;
        _cts.Cancel();
        _writeQueue.CompleteAdding();
        while (_ackWaiters.TryDequeue(out var waiter))
            waiter.TrySetResult(new FileAck { Id = "", Error = "session closed" });
        _peerHello.TrySetCanceled();
        _established.TrySetCanceled();
        try { _closeTransport(); } catch { /* transport already gone */ }
        try { _listener.OnClosed(reason); } catch { /* listener must not throw here */ }
    }

    public void Dispose()
    {
        TearDown("disposed");
        _cts.Dispose();
    }

    // ------------------------------------------------------------- handshake

    private async Task HandshakeAsync()
    {
        try
        {
            if (_role == BdipRole.Initiator)
                await SendAsync(new Frame(Bdip.TypeHello, _hello.Encode()));

            var peerFrame = await _peerHello.Task.WaitAsync(
                TimeSpan.FromMilliseconds(Bdip.HelloTimeoutMs));
            var info = HelloInfo.Decode(peerFrame.Payload);
            if (!string.IsNullOrEmpty(_hello.Token) && !string.IsNullOrEmpty(info.Token)
                && _hello.Token != info.Token)
            {
                Fail("pairing token mismatch");
                return;
            }
            if (_role == BdipRole.Responder)
                await SendAsync(new Frame(Bdip.TypeHello, _hello.Encode()));
            _established.TrySetResult();
            _listener.OnEstablished(info);
        }
        catch (Exception)
        {
            if (!_closed) Fail("hello timeout");
        }
    }

    // ------------------------------------------------------------- loops

    private async Task ReaderLoop()
    {
        try
        {
            while (!_closed)
            {
                var frame = await FrameCodec.ReadAsync(_stream, _cts.Token);
                Interlocked.Exchange(ref _lastReceivedAt, Environment.TickCount64);
                if (!IsEstablished)
                {
                    switch (frame.Type)
                    {
                        case Bdip.TypeHello:
                            _peerHello.TrySetResult(frame);
                            continue;
                        case Bdip.TypeBye:
                            TearDown($"peer bye ({(frame.Payload.Length > 0 ? frame.Payload[0] : 0)})");
                            return;
                        default:
                            Fail($"expected HELLO first, got type 0x{frame.Type:x2}");
                            return;
                    }
                }
                switch (frame.Type)
                {
                    case Bdip.TypeHello:
                        break; // duplicate HELLO after establish: ignore
                    case Bdip.TypePing:
                        await SendAsync(new Frame(Bdip.TypePong, frame.Payload));
                        break;
                    case Bdip.TypeBye:
                        TearDown($"peer bye ({(frame.Payload.Length > 0 ? frame.Payload[0] : 0)})");
                        return;
                    case Bdip.TypeClipboardText:
                        _listener.OnClipboardText(FrameCodec.Utf8(frame.Payload));
                        break;
                    case Bdip.TypeClipboardImage:
                        _listener.OnClipboardImage(frame.Payload);
                        break;
                    case Bdip.TypeFileMeta:
                        _listener.OnFileMeta(FileMeta.Decode(frame.Payload));
                        break;
                    case Bdip.TypeFileChunk when frame.Payload.Length >= 4:
                        var index = BinaryPrimitives.ReadInt32LittleEndian(frame.Payload.AsSpan(0, 4));
                        var data = new byte[frame.Payload.Length - 4];
                        Array.Copy(frame.Payload, 4, data, 0, data.Length);
                        _listener.OnFileChunk(index, data);
                        break;
                    case Bdip.TypeFileAck:
                        var ack = FileAck.Decode(frame.Payload);
                        if (_ackWaiters.TryDequeue(out var waiter))
                            waiter.TrySetResult(ack);
                        _listener.OnFileAck(ack);
                        break;
                    default:
                        break; // unknown type: skip per spec
                }
            }
        }
        catch (ProtocolException ex)
        {
            if (!_closed) Fail(ex.Message);
        }
        catch (Exception ex)
        {
            if (!_closed) TearDown($"link lost: {ex.Message}");
        }
    }

    private async Task WriterLoop()
    {
        try
        {
            foreach (var bytes in _writeQueue.GetConsumingEnumerable(_cts.Token))
            {
                await _stream.WriteAsync(bytes, _cts.Token);
                await _stream.FlushAsync(_cts.Token);
                Interlocked.Exchange(ref _lastSentAt, Environment.TickCount64);
            }
        }
        catch (OperationCanceledException) { /* shutting down */ }
        catch (Exception) { /* reader will notice the dead link */ }
    }

    private async Task HeartbeatLoop()
    {
        try
        {
            while (!_closed)
            {
                await Task.Delay(1_000, _cts.Token);
                var now = Environment.TickCount64;
                if (now - _lastReceivedAt >= Bdip.RxDeadMs)
                {
                    TearDown($"rx dead after {Bdip.RxDeadMs} ms");
                    return;
                }
                if (now - _lastSentAt >= Bdip.PingIntervalMs)
                {
                    await SendAsync(new Frame(Bdip.TypePing, [0, 1, 2, 3, 4, 5, 6, 7]));
                    Interlocked.Exchange(ref _lastSentAt, now);
                }
            }
        }
        catch (OperationCanceledException) { /* shutting down */ }
    }

    // ------------------------------------------------------------- senders

    private Task SendAsync(Frame frame)
    {
        _writeQueue.TryAdd(frame.Encode());
        return Task.CompletedTask;
    }

    public bool SendClipboardText(string text)
    {
        var bytes = FrameCodec.Utf8(text);
        if (bytes.Length > Bdip.MaxText) return false;
        return _writeQueue.TryAdd(new Frame(Bdip.TypeClipboardText, bytes).Encode());
    }

    public bool SendClipboardImage(byte[] png)
    {
        if (png.Length > Bdip.MaxImage) return false;
        return _writeQueue.TryAdd(new Frame(Bdip.TypeClipboardImage, png).Encode());
    }

    public void SendFileAck(FileAck ack) =>
        _writeQueue.TryAdd(new Frame(Bdip.TypeFileAck, ack.Encode()).Encode());

    /// <summary>Streams a file with stop-and-wait chunking (spec §3.4).</summary>
    public async Task<long> SendFileAsync(FileMeta meta, Func<long, int, byte[]?> readChunk, Action<long, long>? onProgress = null)
    {
        await SendAsync(new Frame(Bdip.TypeFileMeta, meta.Encode()));
        long offset = 0;
        uint index = 0;
        while (offset < meta.Size)
        {
            var count = (int)Math.Min(Bdip.ChunkSize, meta.Size - offset);
            var chunk = readChunk(offset, count);
            if (chunk == null || chunk.Length != count) return -1;
            await SendAsync(new Frame(Bdip.TypeFileChunk,
                [.. FrameCodec.U32Le(index), .. chunk]));
            var tcs = new TaskCompletionSource<FileAck>(TaskCreationOptions.RunContinuationsAsynchronously);
            _ackWaiters.Enqueue(tcs);
            var ack = await tcs.Task.WaitAsync(TimeSpan.FromMilliseconds(ChunkAckTimeoutMs));
            if (ack.Error != null || ack.Id != meta.Id) return -1;
            offset += count;
            index++;
            onProgress?.Invoke(offset, meta.Size);
            if (ack.Done == true) return meta.Size; // receiver combined the final ack
        }
        var doneWaiter = new TaskCompletionSource<FileAck>(TaskCreationOptions.RunContinuationsAsynchronously);
        _ackWaiters.Enqueue(doneWaiter);
        var done = await doneWaiter.Task.WaitAsync(TimeSpan.FromMilliseconds(ChunkAckTimeoutMs));
        return done.Done == true && done.Error == null ? meta.Size : -1;
    }
}
